File: 算法/代码随想录算法PDF/00调试源代码/40.组合总和-ii.cpp
/*
 * @lc app=leetcode.cn id=40 lang=cpp
 *
 * [40] 组合总和 II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    void backtracking(vector<int>& candidates, int need, int startIndex) {
        if (need == 0) {
            res.push_back(path);
            return ;
        }
        for (int i = startIndex; i < candidates.size() && need - candidates[i] >= 0; i++) {
            if (i > startIndex && candidates[i] == candidates[i - 1]) continue;  // 同层去重
            path.push_back(candidates[i]);
            backtracking(candidates, need - candidates[i], i + 1);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());  // 排序很重要！
        backtracking(candidates, target, 0);
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/40.组合总和-ii.py
#
# @lc app=leetcode.cn id=40 lang=python3
#
# [40] 组合总和 II
#

# @lc code=start
class Solution:
    def __init__(self):
        self.res = []
        self.path = []
    def backtracking(self, candidates: List[int], need: int, startIndex: int) -> None:
        if need == 0:
            self.res.append(self.path[:])
            return
        for i in range(startIndex, len(candidates)):
            if need - candidates[i] < 0:  # 剪枝
                break
            if i > startIndex and candidates[i] == candidates[i - 1]:  # 去重
                continue
            self.path.append(candidates[i])
            self.backtracking(candidates, need - candidates[i], i + 1)
            self.path.pop()
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        self.backtracking(candidates, target, 0)
        return self.res

# @lc code=end

